Arnold Renderer Backend Features

danieldresser-ie opened the pull request:
Continuing to try and PR clean bits of my baking branch.

Known issues: I wasn't sure where to document the new output parameter - I couldn't find much documentation for the parameters of Outputs.  Also, while putting together a quick unit test, I noticed that I'm not pruning the "camera" parameter before it reaches the backend, so we're ending up with a user parameter named "camera" in the ass file.  This is a bit sloppy, since we handle this parameter ourselves, and I should probably clean it up.

danieldresser-ie commented:
Fixed issue with "camera" parameter being passed to backend, and added additional commit supporting "mesh" parameter for uv_camera.  Only remaining question from my point of view is if I need to document the new features somewhere.

danieldresser-ie commented:
Note that as part of this we should also merge: https://github.com/ImageEngine/cortex/pull/895

danieldresser-ie commented:
I'm ready to put up the final texture bake PR that combines all this stuff togther, and I'm wondering whether we can get some of the prereqs merged first - have you had a chance to look at this one?

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -2980,6 +2988,51 @@ class ArnoldRenderer final : public ArnoldRendererBase
+			for( const auto &it : m_globals->cameraMap() )
+			{
+				IECoreScene::ConstCameraPtr cortexCamera = it.second;
+				auto i = cortexCamera->parameters().find( "mesh" );
```
This could be much simpler. Either use `parametersData()->member<StringData>( "mesh" )` if you genuinely want to silently ignore parameters with the wrong type, or use `parameter( cortexCamera->parameters(), "mesh", "" )` to get reporting for unexpected types - I think the latter is probably better?

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -2616,6 +2616,13 @@ class ArnoldGlobals
 
+		typedef tbb::concurrent_unordered_map<std::string, IECoreScene::ConstCameraPtr> CameraMap;
+
+		const CameraMap &cameraMap()
```
Why does this need to be public now? Can we keep encapsulation by moving `updateCameraMeshes()` to ArnoldGlobals? There are a lot of moving parts in this file, so minimising the coupling between each part is pretty worthwhile...

johnhaddon commented on python/GafferArnoldTest/IECoreArnoldPreviewTest/RendererTest.py:
```diff
@@ -662,6 +663,63 @@ def testOutputLPEs( self ) :
+
+		r.render()
+
+		read = GafferImage.ImageReader()
```
IECoreArnoldPreview shouldn't have Gaffer dependencies - can we use IECoreImage.ImageReader here instead?

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -2532,29 +2538,49 @@ class ArnoldGlobals
-					if( result != AI_SUCCESS )
+					// Loop through all cameras referenced by any current outputs,
+					// and do a render for each
+					std::vector<std::string> cameraOverrides;
```
This needs to be a set! Currently we're doing one render per output, even if they all have the same camera...

danieldresser-ie commented:
OK, I've addressed feedback.  ( Ah, good catch on that duplicated render problem ).

But while testing this a bit more, I realized that I never really properly hammered the very first commit here properly before moving on to texture baking stuff, and I had forgotten that we don't run the Arnold tests on Travis.  Turns out, something is quite badly wrong with that first commit - I've cc'ed John on an email to SolidAngle trying to figure it out.  This is definitely not ready for merging until we get that sorted out.

danieldresser-ie commented:
OK, I've now removed the interactive rendering changes ( pending feedback from SolidAngle ), so this should be OK to merge.  One final check - did you have an issue with how I've organized the adaptive sampling settings in the UI?  I was skimming through the code for a final review, and vaguely recalled something, but I don't see it in writing, so maybe it was a different PR.

johnhaddon commented:
> One final check - did you have an issue with how I've organized the adaptive sampling settings in the UI?

Nope - looks good to me.
